core/src/main/scala/replpp/util/ClasspathHelper.scala

```diff
@@ -0,0 +1,113 @@
+   * the order anyway.
+   */
+  def create(config: Config, quiet: Boolean = false): String = {
+    /** Important: we absolutely have to make sure this starts and ends with a `pathSeparator`.
+     * Otherwise, the last entry is lost somewhere down the line (I didn't find the exact location where things go
+     * wrong, but it looked like somewhere in dotty 3.3.0). */
+    createAsSeq(config, quiet).mkString(pathSeparator, pathSeparator, pathSeparator)
+  }
+
+  protected[util] def createAsSeq(config: Config, quiet: Boolean = false): Seq[Path] = {
+    val entries = Seq.newBuilder[Path]
+    System.getProperty("java.class.path").split(pathSeparator).map(Paths.get(_)).foreach(entries.addOne)
+
+    val fromDependencies = dependencyArtifacts(config)
+    fromDependencies.foreach(entries.addOne)
+    if (fromDependencies.nonEmpty && !quiet) {
+      println(s"resolved dependencies - adding ${fromDependencies.size} artifact(s) to classpath - to list them, enable verbose mode")
+      if (verboseEnabled(config)) fromDependencies.foreach(println)
+    }
+
+    jarsFromClassLoaderRecursively(classOf[replpp.ReplDriver].getClassLoader)
+      .foreach(url => entries.addOne(Paths.get(url.toURI)))
+
+    onlyHighestVersionForEachJar(entries.result().map(versionInfoFromJar)).sorted
+  }
+
+  /**
+   * Goal: for each organisation/name tuple: filter out anything but the highest version. This is to ensure we don't have jars with different versions on the classpath.
+   * Problem: since we don't only get the jars via maven coordinates, but also inherit from existing classpath entries, all we have is a list of jars.
+   * The best solution is probably to use separate classloaders for the REPL itself and the code that's executed within. Similar to what ammonite does..
+   * In the interim, we use this hack as a semi-best-effort basis: group by the information in their manifests. */
+  private[util] def onlyHighestVersionForEachJar(dependencyInfos: Seq[JarWithManifestInfo]): Seq[Path] = {
+    dependencyInfos
+      .groupMap(dep => (dep.organisation, dep.name))(dep => (dep.jar, Version(dep.version.getOrElse("0"))))
+      .map { case ((organisation, name), jars) =>
+        if (jars.size == 1) {
+          jars.head._1
+        } else {
+          // choose the jar with the highest version
+          val result = jars.maxBy { case (_, version) => version } match {
+            case (jar, _) => jar
+          }
+          if (jars.size > 1) {
+            println(s"found ${jars.size} alternatives for organisation=${organisation.getOrElse("")} and name=${name.getOrElse("")}")
+            println(s"-> using the jar with the highest version: $result")
+          }
+          result
+        }
+    }.toSeq
+  }
+
+  @nowarn
+  private def versionInfoFromJar(jar: Path): JarWithManifestInfo = {
+    val manifestBytes = readFileFromZip(jar, "META-INF/MANIFEST.MF")
```

Comment:
do we need any fine-grained error handling here if that manifest file can't be read? We don't if the JVM would just error out loudly later on for missing manifests but I'm not familiar enough with JARs to know whether it's valid for a JARs not to have one.

Reply:
haven't encountered one in my tests, i think that's one of the checks on most public repos. 
either way, good point, I just added some failsafe handling to not crash and burn if the manifest is broken or nonexistent 👍🏻